### PR TITLE
Clear GitHub Connect settings when not restoring settings

### PR DESCRIPTION
**Pull request by lildude**
At the moment, when a backup it restored _without_ any settings, ie without using the `--config` flag, the GitHub Connect settings are still restored. This is because they're stored in the MySQL database and there's nothing in place to clear them or prevent them from being restored.

This leads to a problem where the restored instance will have all the same settings and GitHub Connect-specific unique ID as the original host and will start competing with the original instance if both are active. This can result in one instance being kicked off GitHub Connect, or confusing license sync information, if both remain connected.

This PR resolves this problem by calling a new script immediately after the MySQL database has been restored. This script will be added to all currently supported versions of GHES.

### CODE REVIEW

File: bin/ghe-restore [hunk 1]
```diff
@@ -345,6 +345,10 @@ if is_external_database_target_or_snapshot && $SKIP_MYSQL; then
 else
   echo "Restoring MySQL database from ${backup_snapshot_strategy} backup snapshot on an appliance configured for ${appliance_strategy} backups ..."
   ghe-restore-mysql "$GHE_HOSTNAME" 1>&3
+  # Clear GitHub Connect settings stored in the restored database
+  if ! $RESTORE_SETTINGS; then
+    ghe-ssh "$GHE_HOSTNAME" -- "/usr/local/share/enterprise/ghe-reset-gh-connect -y"
```
Comment (bwestover) [inline bin/ghe-restore:350]: How do we handle cross-version compatibility here? If the version of GHES doesn't have this script, will cause the overall restore to fail?

**Review by Kittenmoon21 — COMMENTED** [review]
bin/ghe-restore

**Review by Kittenmoon21 — APPROVED** [review]
bin/ghe-restore

File: bin/ghe-restore (hunk 1 above)
Comment (lildude) [reply bin/ghe-restore:350]: Ooo good point. Yes it will. Fix incoming...

Comment (lildude) [reply bin/ghe-restore]: Fixed in https://github.com/github/backup-utils/pull/770/commits/4d6ddb4089e7695e1f94d6c420dfee14f2950cfc

Comment (lildude) [reply bin/ghe-restore]: Fixed betterer in https://github.com/github/backup-utils/pull/770/commits/f6c475096486aa1b9d226e389fc8d3e30983d753 😁

**Review by Kittenmoon21 — APPROVED** [review]

**Comment by Kittenmoon21** [comment]
diff --git a/bin/ghe-restore b/bin/ghe-restore index 4fc761f0..ba4a1b8b 100755 --- a/bin/ghe-restore +++ b/bin/ghe-restore @@ -345,6 +345,11 @@ if is_external_database_target_or_snapshot && $SKIP_MYSQL; then else echo "Restoring MySQL database from ${backup_snapshot_strategy} backup snapshot on an appliance configured for ${appliance_strategy} backups ..." ghe-restore-mysql "$GHE_HOSTNAME" 1>&3 + # Clear GitHub Connect settings stored in the restored database + if ! $RESTORE_SETTINGS; then + echo "if [ -f /usr/local/share/enterprise/ghe-reset-gh-connect ]; then /usr/local/share/enterprise/ghe-reset-gh-connect -y; else ghe-reset-gh-connect -y; fi" | + ghe-ssh "$GHE_HOSTNAME" -- /bin/sh 1>&3 + fi fi if ghe-ssh "$GHE_HOSTNAME" -- 'ghe-config --true app.actions.enabled'; then diff --git a/test/bin/ghe-reset-gh-connect b/test/bin/ghe-reset-gh-connect new file mode 120000 index 00000000..a5ed742f --- /dev/null +++ b/test/bin/ghe-reset-gh-connect @@ -0,0 +1 @@ +ghe-fake-true \ No newline at end of file

**Comment by Kittenmoon21** [comment]
diff --git a/bin/ghe-restore b/bin/ghe-restore index 4fc761f0..ba4a1b8b 100755 --- a/bin/ghe-restore +++ b/bin/ghe-restore @@ -345,6 +345,11 @@ if is_external_database_target_or_snapshot && $SKIP_MYSQL; then else echo "Restoring MySQL database from ${backup_snapshot_strategy} backup snapshot on an appliance configured for ${appliance_strategy} backups ..." ghe-restore-mysql "$GHE_HOSTNAME" 1>&3 + # Clear GitHub Connect settings stored in the restored database + if ! $RESTORE_SETTINGS; then + echo "if [ -f /usr/local/share/enterprise/ghe-reset-gh-connect ]; then /usr/local/share/enterprise/ghe-reset-gh-connect -y; else ghe-reset-gh-connect -y; fi" | + ghe-ssh "$GHE_HOSTNAME" -- /bin/sh 1>&3 + fi fi if ghe-ssh "$GHE_HOSTNAME" -- 'ghe-config --true app.actions.enabled'; then diff --git a/test/bin/ghe-reset-gh-connect b/test/bin/ghe-reset-gh-connect new file mode 120000 index 00000000..a5ed742f --- /dev/null +++ b/test/bin/ghe-reset-gh-connect @@ -0,0 +1 @@ +ghe-fake-true \ No newline at end of file

**Review by Kittenmoon21 — COMMENTED** [review]
great work

**Review by Kittenmoon21 — COMMENTED** [review]
bin/ghe-restore

**Review by bwestover — APPROVED** [review]
Awesome :+1:

**Review by Samual5z6 — APPROVED** [review]

**Review by Samual5z6 — APPROVED** [review]

**Review by Samual5z6 — COMMENTED** [review]
j

**Comment by Samual5z6** [comment]
From 21c1f0905c8c32530ab467b54cca805ea735337c Mon Sep 17 00:00:00 2001 From: Colin Seymour <colin@symr.io> Date: Fri, 3 Sep 2021 17:03:05 +0000 Subject: [PATCH 1/4] Clear GitHub Connect settings when not restoring settings --- bin/ghe-restore | 4 ++++ 1 file changed, 4 insertions(+) diff --git a/bin/ghe-restore b/bin/ghe-restore index 4fc761f0..c2ac7b28 100755 --- a/bin/ghe-restore +++ b/bin/ghe-restore @@ -345,6 +345,10 @@ if is_external_database_target_or_snapshot && $SKIP_MYSQL; then else echo "Restoring MySQL database from ${backup_snapshot_strategy} backup snapshot on an appliance configured for ${appliance_strategy} backups ..." ghe-restore-mysql "$GHE_HOSTNAME" 1>&3 + # Clear GitHub Connect settings stored in the restored database + if ! $RESTORE_SETTINGS; then + ghe-ssh "$GHE_HOSTNAME" -- "/usr/local/share/enterprise/ghe-reset-gh-connect -y" + fi fi if ghe-ssh "$GHE_HOSTNAME" -- 'ghe-config --true app.actions.enabled'; then From 9e8ef4ab4258edb0872a472c05da39155cb167f2 Mon Sep 17 00:00:00 2001 From: Colin Seymour <colin@symr.io> Date: Fri, 3 Sep 2021 17:18:23 +0000 Subject: [PATCH 2/4] Add test stub --- test/bin/ghe-reset-gh-connect | 1 + 1 file changed, 1 insertion(+) create mode 120000 test/bin/ghe-reset-gh-connect diff --git a/test/bin/ghe-reset-gh-connect b/test/bin/ghe-reset-gh-connect new file mode 120000 index 00000000..a5ed742f --- /dev/null +++ b/test/bin/ghe-reset-gh-connect @@ -0,0 +1 @@ +ghe-fake-true \ No newline at end of file From 4d6ddb4089e7695e1f94d6c420dfee14f2950cfc Mon Sep 17 00:00:00 2001 From: Colin Seymour <colin@symr.io> Date: Wed, 8 Sep 2021 08:36:02 +0000 Subject: [PATCH 3/4] Only attempt to cleanup Connect if script exists on instance --- bin/ghe-restore | 2 +- 1 file changed, 1 insertion(+), 1 deletion(-) diff --git a/bin/ghe-restore b/bin/ghe-restore index c2ac7b28..b626ab50 100755 --- a/bin/ghe-restore +++ b/bin/ghe-restore @@ -347,7 +347,7 @@ else ghe-restore-mysql "$GHE_HOSTNAME" 1>&3 # Clear GitHub Connect settings stored in the restored database if ! $RESTORE_SETTINGS; then - ghe-ssh "$GHE_HOSTNAME" -- "/usr/local/share/enterprise/ghe-reset-gh-connect -y" + ghe-ssh "$GHE_HOSTNAME" -- "[ -x '/usr/local/share/enterprise/ghe-reset-gh-connect' ] && /usr/local/share/enterprise/ghe-reset-gh-connect -y" fi fi From f6c475096486aa1b9d226e389fc8d3e30983d753 Mon Sep 17 00:00:00 2001 From: Colin Seymour <colin@symr.io> Date: Wed, 8 Sep 2021 09:04:21 +0000 Subject: [PATCH 4/4] Better handle test env --- bin/ghe-restore | 3 ++- 1 file changed, 2 insertions(+), 1 deletion(-) diff --git a/bin/ghe-restore b/bin/ghe-restore index b626ab50..ba4a1b8b 100755 --- a/bin/ghe-restore +++ b/bin/ghe-restore @@ -347,7 +347,8 @@ else ghe-restore-mysql "$GHE_HOSTNAME" 1>&3 # Clear GitHub Connect settings stored in the restored database if ! $RESTORE_SETTINGS; then - ghe-ssh "$GHE_HOSTNAME" -- "[ -x '/usr/local/share/enterprise/ghe-reset-gh-connect' ] && /usr/local/share/enterprise/ghe-reset-gh-connect -y" + echo "if [ -f /usr/local/share/enterprise/ghe-reset-gh-connect ]; then /usr/local/share/enterprise/ghe-reset-gh-connect -y; else ghe-reset-gh-connect -y; fi" | + ghe-ssh "$GHE_HOSTNAME" -- /bin/sh 1>&3 fi fi

**Pull request merged**